### PR TITLE
api: remove waitForLoadState() in favor of PageEvent.page(options)

### PR DESCRIPTION
**Pull request by dgozman**

### CODE REVIEW

**Comment by jperl** [comment]
This is a useful helper for third party libraries, for example we use it in [waitForPage](https://github.com/qawolf/playwright-utils/blob/736aa28cea3bcbd862b910715d27f5f1e9279305/src/context/waitForPage.ts). So it would be nice to be able to rely on something like `waitForLoadState`.

The purpose of waitForPage is to have a standard way of looking up pages based on their creation order. This is useful for recording a test that tests a popup `let popup = await waitForPage(context, 1)`.

File: docs/api.md [hunk 1]
```diff
@@ -1895,8 +1871,27 @@ This method returns all of the dedicated [WebWorkers](https://developer.mozilla.
...
+  - `waitUntil` <"commit"|"load"|"domcontentloaded"|"networkidle0"|"networkidle2"> When to consider the page to be loaded, defaults to `load`. Events can be either:
+    - `'commit'` - navigation is committed, new url is displayed in the browser address bar.
+    - `'load'` - consider navigation to be finished when the `load` event is fired.
+    - `'domcontentloaded'` - consider navigation to be finished when the `DOMContentLoaded` event is fired.
+    - `'networkidle0'` - consider navigation to be finished when there are no more than 0 network connections for at least `500` ms.
+    - `'networkidle2'` - consider navigation to be finished when there are no more than 2 network connections for at least `500` ms.
+- returns: <[Promise]<[Page]>> Promise which resolves to the created page once it loads, according to `waitUntil` option.
+
+This resolves when the page reaches a required load state, `load` by default.
```
Comment (yury-s) [inline docs/api.md]: Let's add an option for 'nowait'

**Review by yury-s — APPROVED** [review]

File: docs/api.md (hunk 1 above)
Comment (pavelfeldman) [reply docs/api.md]: Why? You can't do much to the page that has not committed yet.

**Comment by yury-s** [comment]
> The purpose of waitForPage is to have a standard way of looking up pages based on their creation order. We use it to make creating tests with multiple pages cleaner.
> 
@jperl IIUC you should still be able to set `IndexedPage.createdIndex` correctly. 'page' events will fire in order of the page creation, the new API just forces clients to wait for the `Page` to be initialized and get to a particular state before using it. After the page reached the state you can set the creationIndex:
```js
const creationIndex = ++lastPageCreationIndex;
event.page().then(p => p.creationIndex = creationIndex);
```
. context.pages()` today returns only pages which has already been initialized by playwright but may not have reached the same state as you'd expect in `PageEvent.page()`. One way to address that non-determinism would be to pass similar options to `BrowserContext.pages()`. Would that work for your scenario?

**Comment by jperl** [comment]
Yes if context.pages allowed passing a LifecycleEvent to filter by `pages({ waitUntil: "load" })`, that would solve my problem nicely.

**Comment by jperl** [comment]
We want to mark the creationIndex immediately, in case there is a race. And allow users to decide on the waitUntil lifecycle event in `waitForPage`, which is why `context.pages({ waitUntil })` would be necessary.

I could hack around it by marking a bunch of load states on the page inside the "page" event handler, but that is not as clean as `context.pages({ waitUntil: })` or `page.waitForLoadState`.

**Comment by yury-s** [comment]
Another approach that could work without `context.pages()` is something like this:

```js
const indexedPages = new Set;
const lastPageIndex = 0;
context.on('page', event => {
  const pageIndex = ++lastPageIndex;
  event.page().then(page => {
    page.creationIndex = pageIndex;
    indexedPages.add(page);
  });  
});
```
Now that we have `page` event per context and each context starts empty enables this logic which was impossible before.

**Comment by jperl** [comment]
Yes I already do it this way https://github.com/qawolf/playwright-utils/blob/master/src/context/indexPages.ts the issue is allowing waitForPage to choose which lifecycle event to wait for.

**Comment by dgozman** [comment]
I find api like `waitForPage` hard to use.
- For example, if your client calls `waitForPage` at some random point, it will currently wait for some random navigation to reach `load`, not the initial navigation (as you probably intended?)
- Also, calling `context.pages()` will wait until all pages from the context are initialized, and that could happen long after the page you are waiting for has reached the load state (or even, redirected away to somewhere else).

I'd suggest to use `PageEvent` to assign creation index, and explicitly call `event.page(options)` with desired `waitUntil`.

**Comment by jperl** [comment]
> For example, if your client calls waitForPage at some random point, it will currently wait for some random navigation to reach load, not the initial navigation (as you probably intended?)

The point of the `waitForPage` API is to make generating tests for multiple pages easier. The `waitForPage` is called right before the first action on that page. Ex:

```
test("pscp", async () => {
  await page.goto("https://pscp.tv/");
  await page.click("[data-qa=login]");
  // switch to twitter popup
  page = await qawolf.waitForPage(page.context(), 1);
  await page.type("[data-qa=username]", "USERNAME");
  await page.type("[data-qa=password]", "PASSWORD");
  await page.press("[data-qa=login]", "Enter");
  // switch back to application
  page = await qawolf.waitForPage(page.context(), 0);
  await page.click("[data-qa=logout]");
});
```

> Also, calling context.pages() will wait until all pages from the context are initialized, and that could happen long after the page you are waiting for has reached the load state (or even, redirected away to somewhere else).

Good point, so I would prefer `waitForLoadState`. That being said, I can workaround this.

File: docs/api.md (hunk 1 above)
Comment (yury-s) [reply docs/api.md]: Well, you cannot get not-committed page here anyway, if 'commit' is the earliest state the page can be in here I'm fine with that but should articulate that in the documentation.

File: docs/api.md [hunk 2]
```diff
@@ -1895,8 +1871,29 @@ This method returns all of the dedicated [WebWorkers](https://developer.mozilla.
...
+
+```js
+const [popup] = await Promise.all([
```
Comment (yury-s) [inline docs/api.md]: `[,popup]`

File: src/page.ts [hunk 3]
```diff
@@ -88,20 +89,56 @@ export type FileChooser = {
...
+      waitUntil = 'load',
+    } = options;
+    const lifecyclePromise = this._lifecyclePromises.get(waitUntil);
+    if (!lifecyclePromise)
+      throw new Error(`Unsupported waitUntil option ${String(waitUntil)}`);
+
+    const promises = [ lifecyclePromise ];
+    let timer: NodeJS.Timer | undefined;
+    if (timeout) {
+      promises.push(new Promise(f => timer = setTimeout(() => {
```
Comment (yury-s) [inline src/page.ts]: Use `waitWithTimeout` ?

**Review by yury-s — APPROVED** [review]

**Pull request merged**